relay-server/src/processing/transactions/mod.rs

```diff
@@ -375,27 +450,98 @@ impl<T: Counted + AsRef<Annotated<Event>>> RateLimited for Managed<ExpandedTrans
+        debug_assert!(!self.flags.spans_extracted);
+
         let ExpandedTransaction {
             headers: _,
-            transaction,
+            event,
             flags,
             attachments,
             profile,
             extracted_spans,
+            category: _,
         } = self.as_ref();
 
         // If there is a transaction limit, drop everything.
         // This also affects profiles that lost their transaction due to sampling.
-        for (category, quantity) in transaction.quantities() {
+        let limits = rate_limiter
+            .try_consume(scoping.item(DataCategory::Transaction), 1)
+            .await;
+
+        // We do not check indexed quota at this point, because metrics have not been extracted
+        // from the transaction yet.
+
+        let attachment_quantities = attachments.quantities();
+
+        // Check profile limits:
+        for (category, quantity) in profile.quantities() {
             let limits = rate_limiter
                 .try_consume(scoping.item(category), quantity)
                 .await;
 
             if !limits.is_empty() {
-                return Err(self.reject_err(Error::from(limits)));
+                self.modify(|this, record_keeper| {
+                    record_keeper.reject_err(Error::from(limits), this.profile.take());
+                });
             }
         }
 
+        // Check attachment limits:
+        for (category, quantity) in attachment_quantities {
+            let limits = rate_limiter
+                .try_consume(scoping.item(category), quantity)
+                .await;
+
+            if !limits.is_empty() {
+                self.modify(|this, record_keeper| {
+                    record_keeper
+                        .reject_err(Error::from(limits), std::mem::take(&mut this.attachments));
+                });
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl RateLimited for Managed<ExpandedTransaction<Indexed>> {
+    type Error = Error;
+
+    async fn enforce<R>(
+        &mut self,
+        mut rate_limiter: R,
+        ctx: Context<'_>,
+    ) -> Result<(), Rejected<Self::Error>>
+    where
+        R: RateLimiter,
+    {
+        let scoping = self.scoping();
+
+        // These debug assertions are here because this function does not check indexed or extracted
+        // span limits.
+        // TODO: encode flags into types instead.
+        debug_assert!(self.flags.metrics_extracted);
+
+        let ExpandedTransaction {
+            headers: _,
+            event,
+            flags,
+            attachments,
+            profile,
+            extracted_spans,
+            category: _,
+        } = self.as_ref();
+
+        // If there is a transaction limit, drop everything.
+        // This also affects profiles that lost their transaction due to sampling.
+        let limits = rate_limiter
+            .try_consume(scoping.item(DataCategory::TransactionIndexed), 1)
```

Comment:
I think you need to check both categories here, that's what the `EnvelopeLimiter` [does](https://github.com/getsentry/relay/blob/aad4b6099d12422e88dd5df49abae11247efdd99/relay-server/src/utils/rate_limits.rs#L693-L702).

Note `fn is_event_active()` checks both categories, indexed and non-indexed as well.